### PR TITLE
EN-13432: Nodes shuffler refactoring

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
## Reasoning behind the pull request
- nodes shuffler used static values for min nodes per shard / min nodes in meta / hysteresis / adaptivity
  
## Proposed changes
- integrate `chainParametersHandler` into it

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/ElrondNetwork/elrond-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: sharding/nodesCoordinator/hashValidatorShuffler.go [hunk 1]
```diff
@@ -18,13 +18,10 @@ var _ NodesShuffler = (*randHashShuffler)(nil)
...
-	NodesMeta            uint32
-	Hysteresis           float32
-	Adaptivity           bool
-	ShuffleBetweenShards bool
-	MaxNodesEnableConfig []config.MaxNodesChangeConfig
-	EnableEpochsHandler  common.EnableEpochsHandler
+	ShuffleBetweenShards   bool
+	MaxNodesEnableConfig   []config.MaxNodesChangeConfig
+	EnableEpochsHandler    common.EnableEpochsHandler
+	ChainParametersHandler ChainParametersHandler
```
Comment (AdoAdoAdo) [inline sharding/nodesCoordinator/hashValidatorShuffler.go:24]: check also if the notify order for chain parameters handler.
it needs to at least be before nodes coordinator.

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: sharding/nodesCoordinator/hashValidatorShuffler.go [hunk 2]
```diff
@@ -49,11 +46,9 @@ type randHashShuffler struct {
 	// when reinitialization of node in new shard is implemented
 	shuffleBetweenShards bool
 
-	adaptivity              bool
-	nodesShard              uint32
-	nodesMeta               uint32
-	shardHysteresis         uint32
-	metaHysteresis          uint32
+	currentChainParameters config.ChainParametersByEpochConfig
+	//shardHysteresis         uint32
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply sharding/nodesCoordinator/hashValidatorShuffler.go:50]: removed

File: sharding/chainParametersHolder.go [hunk 3]
```diff
@@ -39,24 +41,19 @@ func NewChainParametersHolder(args ArgsChainParametersHolder) (*chainParametersH
...
 	}
 
-	currentParams, err := getMatchingChainParametersUnprotected(args.EpochNotifier.CurrentEpoch(), args.ChainParameters)
-	if err != nil {
-		return nil, err
-	}
-
 	paramsHolder := &chainParametersHolder{
-		currentChainParameters: currentParams,
+		currentChainParameters: earliestChainParams, // will be updated on the epoch notifier handlers
```
Comment (iulianpascalau) [inline sharding/chainParametersHolder.go:45]: yup, that is the behavior right now. I do not see a good reason why we should change this in the future.

**Review by iulianpascalau — APPROVED** [review]

File: sharding/chainParametersHolder_test.go [hunk 4]
```diff
@@ -6,8 +6,9 @@ import (
 	"testing"
 
 	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/data/block"
 	"github.com/ElrondNetwork/elrond-go/config"
-	"github.com/ElrondNetwork/elrond-go/testscommon/epochNotifier"
+	"github.com/ElrondNetwork/elrond-go/sharding/mock"
```
Comment (AdoAdoAdo) [inline sharding/chainParametersHolder_test.go:11]: can you use one epoch start notifier from testscommon?
so we can gradually remove all other mocks

Comment (bogdan-rosianu) [reply sharding/chainParametersHolder_test.go]: done + removed other unused mocks and stubs

**Review by AdoAdoAdo — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**